proposals/distribution.md

```diff
@@ -0,0 +1,66 @@
+
+In the past when the topic of having an OCI specification around the distribution of container images was discussed, it was deferred as "let’s get the image format defined, meanwhile the industry will settle on a distribution standard". Fast forward, OCI image format is out and adopted, and the Registry v2 is the defacto standard. There is and will be use-cases for alternate methods and the future will likely hold creative ways to push, fetch and share container images, but right now this promotion serves to acknowledge by the OCI the current industry standard of distributing container images.
+
+There is polish that is needed e.g. broken links to storage-driver docs, as well as making sections more generic regarding the OCI descriptors and media-types, but on the whole this is a lateral move.
+
+## Initial Maintainers
+
+* Stephen Day <stephen.day@docker.com> (@stevvooe)
+* Vincent Batts <vbatts@redhat.com> (@vbatts)
+* Derek McGowan <derek.mcgowan@docker.com> (@dmcgowan)
+
+Additional Maintainers to consider:
+
+* Ahmet Alp Balkan (Google)
+* Matt Moore (Google)
+* Yuwa (MSFT)
+* Clayton Coleman (Red Hat)
+* Antonio Murdaca (@runcom) (Red Hat)
+* Samuel Karp (@samuelkarp) (AWS)
+* Mike Brown (IBM)
+* Jimmy Zelinskie jimmy@coreos.com (@jzelinskie)
+* Liu Genping <[liugenping@huawei.com](mailto:liugenping@huawei.com)>
+* Vanessa Sochat (@vsoch) (Stanford) <vsochat@stanford.edu>
+* Eduardo Arango (@ArangoGutierrez) (Sylabs) <eduardo@sylabs.io>
+
+## Code of Conduct
+
+This project would incorporate (by reference) the OCI Code of Conduct ([https://github.com/opencontainers/tob/blob/master/code-of-conduct.md](https://github.com/opencontainers/tob/blob/master/code-of-conduct.md)).
+
+## Governance and Releases
+
+This project would incorporate the Governance and Releases processes from the OCI project template: [https://github.com/opencontainers/project-template](https://github.com/opencontainers/project-template).
+
+## Project Communications
+
+Both of the proposed projects would continue to use existing channels in use by the OCI developer community for communication including:
+
+* GitHub for issues and pull requests
+* The dev@opencontainers.org email list
+* The monthly OCI developer community conference call
+* The #OpenContainers freenode IRC channel
+
+## Versioning / Roadmap
+
+The API spec is currently considered v2 and we will start the specification at v2.0. Fewer places to change and compare, and it would keep with it being a lateral move.
+
+## Frequently Asked Questions (FAQ)
+
+* Does this include the code of the docker-registry?
+    * No. This is an API specification discussion.
+
+## Related GitHub Issues
```

Comment:
What about the proposal of new APIs to the spec? Shall we do it after the vote as future incremental improvements, or shall we raise and discuss them now?

From our experience of running our registry service (Azure Container Registry), there are quite some popular feedbacks/lessons we learned and would like to see if they could be included in the future registry spec:
1. Richer image management ability, including delete repository, purge unreferenced blobs, etc.
2. Richer discovery ability where user has lots of images, like list tag by creation time, list manifest by creation time, etc. So basically to support more query parameters in the list APIs.
3. Richer metadata APIs. Basically user wants to put additional metadata with their image, in additional to a simple tag.
4. Native multi-tenant support. Something like /v2/tenants/{tenant}/repositories/{name}/manifests/{reference}. Since more and more public cloud services are adopting the registry service, a native support for multi-tenant would be great IMHO.

We would like to learn if there is any opportunity that these feedback could be included in the distribution spec. Thanks!

Reply:
Image size and format would also be very useful, beyond sniffing with a `HEAD` request. There is also a distinction between a tag (e.g., latest) and a version (a hash or commit), we've had users asking for both with Singularity Hub/Registry, and then for all these requests, the default should be reasonable of course! Are we going with docker defaults like `latest` for tag, `library` for namespace, and other registries should follow suit? @yuwaMSFT2 for tenant would that be akin to the manifests list where you can request a particular OS or architecture?  The user client would then always be required to make two calls https://github.com/docker/distribution/blob/master/docs/spec/manifest-v2-2.md#manifest-list (and this is what we are doing currently in Singularity)!

Reply:
@yuwaMSFT2 @vsoch interesting extensions. See above proposal to add a sentence about discussing and scheduling in extensions without going into specific detail.

Reply:
@vsoch no the metadata of the manifest is not what I meant for tenant.
It's more to make the repository hierarchical. Like on dockerhub you can have organization, then under the organization you can have repository. Current uri would be /v2/orgname/reponame/, and the registry treat the whole path component "/orgname/reponame" as name parameter in the route matching. It will be good if we can separate it.

Reply:
@mikebrow  this looks good to me!
@yuwaMSFT2 this is badly needed, and I strongly +1. Right now we clump the entire thing that comes before the digest (after the @) and tag (after the :) and the image name (e.g., ubuntu) as the namespace. But given registry urls and the potential for local registries to decide to use a custom namespace, parsing this string has been [more than challenging](https://github.com/singularityware/singularity/blob/master/libexec/python/shell.py#L82)). Being able to separate these two things? :pray: !!

Reply:
@yuwaMSFT2 A lot of the "extensions" you've referenced are really about content management, rather than content distribution. This API has almost always been about content distribution. Adding these extensions to the specification provides little room for vendors to implement these kinds of functionality in a way that fits their platform.

> Right now we clump the entire thing that comes before the digest (after the @) and tag (after the :) and the image name (e.g., ubuntu) as the namespace.

@vsoch None of this is a part of this specification. Even the hub behavior @yuwaMSFT2 was referencing is the hubs implementation. This specification allows arbitrary paths. Other things, like `latest`, are completely up to the client implementation and the registry doesn't care and this specification doesn't even mention.

Let's make sure this exercise is critiquing the specification and not Docker's implementation of it.

Reply:
@yuwaMSFT2 agreed about content management vs. distribution, I've never distinguished those two before.

In that the uri is serving as an entry point into a registry, even if it doesn't conform to a specific namespace, tag, arguably I should be able to (knowing the API conforms to the registry) programatically predict the base uri to be calling based on this alone. It could be as general as a general flow of regular expressions to follow, or as specific as a discrete set of different kinds (e.g., abstract uri vs. an arbitrary path). It would be a stronger specification to have some level of predictibility here, otherwise I still need to write a custom thing (after checking!) for each registry endpoint to even interact with most endpoints.

Reply:
> agreed about content management vs. distribution, I've never distinguished those two before.

The two systems have vastly different requirements. For example, even though the Hub looks like its the same system, they are logically separate between the registry and the hub ui, in practice. This prevents the registry deployment from getting complicated. All of the UI and management of images gets implemented in a separate system, allowing them to grow in features and functionality.

I'm still not quite understanding what you're talking about regarding URIs: the format that you're talking about parsing has nothing to do with this specification. Different systems may implement completely different behaviors and formats, depending on their opinion.

Reply:
@vsoch @yuwaMSFT2 I expanded a bit on the divide between content distribution and management in https://github.com/opencontainers/tob/pull/37#issuecomment-360908827. I'll repost here, as there may have been some confusion.

This API had everything in it required to integrate with a container runtime (ie. pull an image, figure out what tags are in a repo, etc.) and APIs required to integrate with content management systems. This includes listing repositories and tags. There is also a notification API that enables in the registry implementation, but isn't a part of this specification (there may be an argument about pulling this in).